lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -725,6 +731,96 @@ struct ConvertLayoutOpConversion
     return success();
   }
 
+  Value computeStMatrixAddr(Value laneId, int matStride, Location loc,
+                            ConversionPatternRewriter &rewriter) const {
+    Value rowInMat = urem(laneId, i32_val(8)); // row in the 8x8 matrix
+    Value matIndex = udiv(
+        laneId, i32_val(8)); // linear index of the matrix in the 2x2 matrices
+    // Decompose matIndex => s_0, s_1, that is the coordinate in 2x2 matrices in
+    // a warp
+    Value s0 = urem(matIndex, i32_val(2));
+    Value s1 = udiv(matIndex, i32_val(2));
+    Value mIndex = add(rowInMat, mul(s0, i32_val(8)));
+    int m8n8Stride = 8;
+    Value offsetInBytes =
+        add(mul(mIndex, i32_val(matStride)), mul(s1, i32_val(m8n8Stride)));
+    return offsetInBytes;
+  }
+
+  void stMatrixm8n8x4(Value offset, ArrayRef<Value> vals, int indexOffset,
+                      Value smemBase, Type elemTy, Location loc,
+                      ConversionPatternRewriter &rewriter) const {
+    SmallVector<Value> inputs;
+    auto prTy = ptr_ty(rewriter.getContext(), 3);
+    // Pack the input into 2xf16
+    Type packedTy = vec_ty(vals[0].getType(), 2);
+    for (int i = 0; i < 4; i++) {
+      Value input = undef(packedTy);
+      for (int j = 0; j < 2; j++) {
+        input = insert_element(packedTy, input, vals[indexOffset + i * 2 + j],
+                               i32_val(j));
+      }
+      inputs.push_back(bitcast(input, i32_ty));
+    }
+    Value addr = gep(smemBase.getType(),
+                     getTypeConverter()->convertType(elemTy), smemBase, offset);
+    rewriter.create<triton::nvgpu::StoreMatrixOp>(loc, addr, inputs);
+  }
+
+  void storeDistributedToSharedWithStMatrix(
+      RankedTensorType tensorTy, Value llvmSrc, Value smemBase,
+      ArrayRef<unsigned> paddedRepShape, ArrayRef<unsigned> origRepShape,
+      Location loc, ConversionPatternRewriter &rewriter) const {
+    auto shapePerCTA = getShapePerCTA(tensorTy);
+    auto mmaLayout = tensorTy.getEncoding().cast<NvidiaMmaEncodingAttr>();
+    auto order = triton::gpu::getOrder(mmaLayout);
+    auto warpsPerCTA = mmaLayout.getWarpsPerCTA();
+
+    // 4xm8n8 matches exactly the size of 1 warp of wgmma layout for 16bit type
+    // and has a shape of 16x16.
+    int instrN = 16 * warpsPerCTA[order[0]];
```

Comment:
are these supposed to be flipped ? `int instrM = 16 * warpsPerCTA[order[0]];`

Reply:
yeah that was completely off. I reworked the logic